DragaliaAPI.Database/Repositories/AbilityCrestRepository.cs

```diff
@@ -0,0 +1,51 @@
+﻿using DragaliaAPI.Database.Entities;
+using DragaliaAPI.Shared.Definitions.Enums;
+using DragaliaAPI.Shared.MasterAsset.Models;
+using DragaliaAPI.Shared.PlayerDetails;
+using Microsoft.Extensions.Logging;
+
+namespace DragaliaAPI.Database.Repositories;
+
+internal class AbilityCrestRepository : IAbilityCrestRepository
+{
+    private readonly ApiContext apiContext;
+    private readonly IPlayerDetailsService playerDetailsService;
+    private readonly ILogger<AbilityCrestRepository> logger;
+
+    public AbilityCrestRepository(
+        ApiContext apiContext,
+        IPlayerDetailsService playerDetailsService,
+        ILogger<AbilityCrestRepository> logger
+    )
+    {
+        this.apiContext = apiContext;
+        this.playerDetailsService = playerDetailsService;
+        this.logger = logger;
+    }
+
+    public IQueryable<DbAbilityCrest> AbilityCrests =>
+        this.apiContext.PlayerAbilityCrests.Where(
+            x => x.DeviceAccountId == this.playerDetailsService.AccountId
+        );
+
+    public async Task Add(AbilityCrests abilityCrestId)
+    {
+        this.logger.LogDebug("Adding ability crest {print}", abilityCrestId);
+
+        if (await this.FindAsync(abilityCrestId) is not null)
+        {
+            this.logger.LogDebug("Ability crest was already owned.");
```

Comment:
Could you make this a warning?

I would think in most cases when we want to add ability crests, we should be doing this check higher up so that we can handle the crest already existing and give a substitute, e.g. for missions and quests you get a token amount of eldwater instead.

Reply:
Do you mean like LogError instead of LogDebug? Then ye ofc, and also in AbilityCrestService already added the functionality for returning eldwater/rupies if the print is already owned with AddOrRefund ^^

Reply:
Ohhh wait there's a LogWarning too icic